docs/execution.md

```diff
@@ -0,0 +1,208 @@
+# Smart contracts
+
+## Smart Contract Engine
+
+### Rationale
+
+The smart contract engine (SCE) is decoupled from the consensus system (CSS); they maintain separate ledgers: the SCE ledger and the CSS ledger.
+
+The CSS informs the SCE whenever a block becomes CSS-final, and whenever the blockclique changes in any way.
+
+The SCE executes the blockclique operations after sorting them by increasing block slot: this provides an absolute order of execution.
+
+A special feedback loop is required to send coins from the SCE to the CSS.
+
+A unit of "gas" represents a unit of computational cost. Different instructions might cost different amounts of gas.
+
+### The ExecuteSC operation
+
+The following new operation type is defined:
+
+```rust
+OperationType::ExecuteSC {
+    max_gas: u32,
+    gas_price: Amount,
+    coins: Amount,
+    bytecode: Vec<u8>,
+}
+```
+
+Where:
+* `bytecode` is smart contract bytecode that is understood by the SCE but not the CSS
+* `max_gas` is the maximum amount of gas that the execution of `bytecode` is allowed to cost
+* `gas_price` is the price per unit of gas that the caller is willing to pay for the execution
+* `coins` are extra coins that are spent by tge CSS and are available in the execution context of `bytecode`
+
+Blocks that include ExecuteSC operations must respect the following criteria, or else they are invalid:
+* ExecuteSC spends `max_gas * gas_price + coins` from the sender address in the CSS ledger, which must therefore be available
+* all ExecuteSC operations included in the block must be ordered by decreasing `gas_price`
```

Comment:
Why not by gas_price * max_gas ?

Reply:
> This is spend immediately in CSS coins ?

yes, `max_gas * gas_price + coins` is spent in CSS



> Why not by gas_price * max_gas ?


Actually by decreasing `(gas_price * max_gas) / op_size`